### PR TITLE
add errorhint for nonexisting fields and properties

### PR DESCRIPTION
**Pull request by aplavin**
I played a bit with error hints and crafted this:
```julia
julia> (1+2im).real
ERROR: FieldError: type Complex has no field real, available fields: `re`, `im`

julia> nothing.xy
ERROR: FieldError: type Nothing has no field xy; Nothing has no fields at all.

julia> svd(rand(2,2)).VV
ERROR: FieldError: type SVD has no field VV, available fields: `U`, `S`, `Vt`
Available properties: `V`
```
Implementation is just a few lines. To try out, simply copy & paste 
```julia
Base.Experimental.register_error_hint(FieldError) do io, exc
    fields = fieldnames(exc.type)
    if isempty(fields)
        print(io, "; $(nameof(exc.type)) has no fields at all.")
    else
        print(io, ", available fields: $(join(map(k -> "`$k`", fields), ", "))")
    end
    props = _propertynames_bytype(exc.type)
    isnothing(props) && return
    props = setdiff(props, fields)
    isempty(props) && return
    print(io, "\nAvailable properties: $(join(map(k -> "`$k`", props), ", "))")
end

_extract_val(::Type{Val{V}}) where {V} = V
_extract_val(::Type{Val}) = nothing
_propertynames_bytype(::Type{T}) where {T} = Base.promote_op(Val∘propertynames, T) |> _extract_val
```
to your Julia nightly REPL.

Let me know if such an addition would be welcome to Base Julia!

### CODE REVIEW

**Comment by arhik** [comment]
@aplavin Yes we would want something like that. @tecosaur has tried to bring that up here https://github.com/JuliaLang/julia/pull/54504#issuecomment-2159639207

I suggested a seperate `PropertyError` here https://github.com/JuliaLang/julia/pull/54504#issuecomment-2160935942

requesting @LilithHafner @tecosaur for their opinion.

File: base/errorshow.jl
Comment (arhik) [inline base/errorshow.jl:1121]: _propertynames_bytype(::Type{T}) where {T} = promote_op(Val∘Base.Fix2(propertynames, true), T) |> _extract_val

Comment (arhik) [reply base/errorshow.jl:1121]: I suggested this after approval for #55169. There is reconsideration now. So you can ignore this.

**Comment by arhik** [comment]
@aplavin This PR is practically useful. If you want to go ahead with tests too you can follow this PR #54504.

**Comment by tecosaur** [comment]
I've got very mixed feelings on this PR.

For starters, I'd like to say I'm very interested in improvements to error messages, as this PR does :heart_eyes:.

However, a personal goal I have for 1.12 is to make some progress with https://public.tecosaur.net/JuliaErrors.html (both with the proposal itself, and then implementation), and that would supersede this work.

This makes me wonder if work on errors in 1.12 might be better spent working towards something like my proposal, _then_ doing a push to implement nicer error messages with that. If I can interest anyone in collaborating with me on this, that would be fantastic.

This is just my 2c given the work I plan on doing though.

**Comment by nsajko** [comment]
This PR breaks doctests.

**Comment by nsajko** [comment]
@tecosaur TBH it's not clear what you want to do regarding error messages. How would what you want to do supersede this PR?

**Comment by tecosaur** [comment]
From when I raised this in triage, it was suggested I could look at creating a new "detailed error messages" stdlib with JuliaSyntaxHighlighting and StyledStrings as deps, that implements fancy `showerror` methods for custom error types from `Base` (along with a handful of helpful utility functions for producing nice error messages, like fuzzy string matching and printing a syntax-highlighted line of code from a stacktrace) and can be loaded as part of the default sysimage.

Nothing has actually started happening in this direction yet, as I think more discussion is needed. As you mention, there are some gaps in the writing I've currently put up as well.

**Comment by nsajko** [comment]
The doctests have their own job, "doctest".

**Comment by LilithHafner** [comment]
Re: @tecosaur's plans, this PR does not interfere with those plans, so those plans don't block this PR.

---

Your property name recovery from the type is clever! This sort of "best effort" task is one of the few valid usages of inference results such as via `promote_op` :)

---

Maybe we should add (in this PR or a followup) a `PropertyError` type to avoid the following. And/or document that `FieldError`s are only for literal field access errors.

```julia-repl
julia> struct NamedInts{N}
           names::NTuple{N, Symbol}
           vals::NTuple{N, Int}
       end

julia> Base.propertynames(n::NamedInts) = getfield(n, :names)

julia> function Base.getproperty(n::NamedInts, name::Symbol)
           i = findfirst(==(name), getfield(n, :names))
           i === nothing && throw(FieldError(typeof(n), name))
           getfield(n, :vals)[i]
       end

julia> ni = NamedInts((:lo, :hi), (1, 2))
NamedInts{2}((:lo, :hi), (1, 2))

julia> ni.lo
1

julia> ni.names
ERROR: FieldError: type NamedInts has no field names
Fields of NamedInts:
names, vals

Stacktrace:
 [1] getproperty(n::NamedInts{2}, name::Symbol)
   @ Main ./REPL[19]:3
 [2] top-level scope
   @ REPL[22]:1
```

---

> I didn't go with test on purpose – would like to hear some confirmation that this is a mergeable PR. Then if design is ok, I can of course add tests.

Yes, this is a mergeable PR.

**Comment by KristofferC** [comment]
> However, a personal goal I have for 1.12 is to make some progress with https://public.tecosaur.net/JuliaErrors.html (both with the proposal itself, and then implementation), and that would supersede this work.

> This makes me wonder if work on errors in 1.12 might be better spent working towards something like my proposal, then doing a push to implement nicer error messages with that. If I can interest anyone in collaborating with me on this, that would be fantastic.

This is a bit too much cookie licking in my opinion. Incremental improvements are always desired.

**Comment by KristofferC** [comment]
Some nitpicking, I don't think the newline after the colon is needed, in fact I think this can be written as

```
ERROR: FieldError: type Complex has no field real, available fields: re, im
```

Also, it might be useful to code quote the field names in the same spirit as https://github.com/JuliaLang/julia/pull/47077

```
ERROR: FieldError: type Complex has no field `real`, available fields: `re`, `im`
```

File: base/errorshow.jl
Comment (KristofferC) [inline base/errorshow.jl:1112]: How can you get the properties of an object when you only have a type?

**Comment by arhik** [comment]
raised separate `PropertyError` issue to keep this PR simple. 

@aplavin We might have to let go that slick `promote_op` and Val trick to retrieve properties from Type. I learned something new that I didn't think is possible from this PR. Thanks.

**Comment by LilithHafner** [comment]
There's nothing wrong with the `promote_op` and `Val` trick in this context. The reason it is often not a good idea is that compiler internals can change the result, but this is just a hint so that is okay.

File: base/errorshow.jl
Comment (LilithHafner) [inline base/errorshow.jl:1111]: We should special case the empty case:

```julia-repl
julia> nothing.bar
ERROR: FieldError: type Nothing has no field `bar`, available fields: 

Stacktrace:
 [1] getproperty(x::Nothing, f::Symbol)
   @ Base ./Base.jl:49
 [2] top-level scope
   @ REPL[1]:1
```

```suggestion
    fields = fieldnames(exc.type)
    isempty(fields)
        print(io, "; $(exc.type) has no fields.")
    end
        print(io, ", available fields:\n$(join(map(k -> "`$k`", fields), ", "))")
    end
```

Also, rather than printing a blank line after the hint, I'd put the field names on the next line.

File: base/errorshow.jl
Comment (vtjnash) [inline base/errorshow.jl:1125]: This _extract_val implementation is unsound in several ways, so if we are going to use this implementation, we must fix that:
```suggestion
function _propertynames_bytype(T::Type)
    inferred_names = promote_op(Val∘propertynames, T)
    inferred_names isa DataType && inferred_names <: Val || return nothing
    inferred_names = inferred_names.parameters[1]
    inferred_names isa NTuple{Symbol} || return nothing
    return Symbol[inferred_names[i] for i in 1:length(inferred_names)]
end
```

Comment (vtjnash) [reply base/errorshow.jl:1125]: We can also make this much more efficient by adding an early exit to catch most cases
```julia
    which(propertynames, (T,)) === which(propertynames, (Any,)) && return nothing
```

Comment (LilithHafner) [reply base/errorshow.jl:1125]: @vtjnash, when you object to an implementation, I would prefer if you offer _why_ it is wrong, ideally with examples (e.g. a type where `_propertynames_bytype` throws) in addition to the proposed fix you've given. That way folks understand what is wrong with the original and can learn to be more valuable contributors.

In this case, I'm guessing the problem is that `promote_op` may return any type wider than the true output. For example, `Union{Val, MyUnrelatedType}`.

Would this be a valid fix?

```suggestion
_extract_val(::Type{Val{V}}) where {V} = V
_extract_val(_) = nothing
_propertynames_bytype(::Type{T}) where {T} = promote_op(Val∘propertynames, T) |> _extract_val
```

Comment (ararslan) [reply base/errorshow.jl:1125]: > Not sure what "unsound" means in this context

I think Jameson means that it's unsound in the [logic sense](https://en.wikipedia.org/wiki/Soundness): either it's invalid or it has at least one false premise. In this case I assume Jameson means the latter, i.e. that it makes some assumptions that may not hold in practice.

Since `propertynames` is a generic function, a user may add a method to `propertynames` for objects with a particular type. However, there is no guarantee that what the method returns is consistent with what's expected. As an example, say someone defined `propertynames(::CoolType)` to return a `Vector{Symbol}` rather than a tuple of `Symbol`s. Similarly, `Val(::CoolType) isa Val` may not hold, or the result of `promote_op` may not be a `Val`, as Lilith mentioned. (In fact, even `|>` is a generic function.) Thus, in Jameson's suggestion, he's explicitly checking that the result of the `promote_op` call is in fact a `Val{NTuple{Symbol}}`. These kinds of checks are helpful e.g. to ensure that method definitions don't have unexpected downstream consequences.

> Would this be a valid fix?

My Jameson hat unfortunately does not grant me his omniscience but I think what you suggest would be mostly equivalent to his suggestion with the addition of a type assertion upon unwrapping the `Val`? Like:
```julia
_extract_val(::Type{Val{V}}) where {V} = V::NTuple{Symbol}
```

Comment (nsajko) [reply base/errorshow.jl:1125]: Another benefit of the diff proposed by @vtjnash is less specialization.

Comment (LilithHafner) [reply base/errorshow.jl:1125]: > with the addition of a type assertion upon unwrapping the `Val`?

I intentionally omitted that because if someone returns an object that quacks like an `NTuple{N, Symbol}` (i.e. supports `isempty`, `map`, and interpolation of its elements) then we should use it.

---

> Another benefit of the diff proposed by @vtjnash is less specialization.

IMO dispatch is a much more elegant way to extract information from types than inspecting type internals.

> Is Base really written robustly to overloads of such fundamental methods? That's honestly impressive.

No, but we do try when it's not too much trouble.

**Comment by tecosaur** [comment]
Just for fun, I'll mention that I've recently added something similar to my feature branch, but I'm not doing anything with properties at all:

image

Along the way, I've ended up wondering if to resolve the field/property ambiguity whether or not we could introduce a `PropertyError` to distinguish between the two cases without guessing? If we do that, we could have the property error take this form instead:

```julia
struct PropertyError
    type::Type
    property::Symbol
    properties::Vector{Symbol}
end
```

I'd think we could run `propertynames(x)` on the object when constructing this error.

As a result, we'd be able to produce decent error messages with `DataFrame`s and similar types where statically guessing the properties won't turn out well.

**Comment by tecosaur** [comment]
> Great if this is somehow possible without requiring opt-in from all types with getproperty! Not sure if it is though...

It might need types with a custom `getproperty` method to manually throw a `PropertyError`. See the stacktrace we get from `DataFrames`:

```julia-repl
julia> d.foobar
ERROR: ArgumentError: column name :foobar not found in the data frame
Stacktrace:
 [1] lookupname
   @ ~/.julia/packages/DataFrames/58MUJ/src/other/index.jl:413 [inlined]
 [2] getindex
   @ ~/.julia/packages/DataFrames/58MUJ/src/other/index.jl:422 [inlined]
 [3] getindex(df::DataFrame, ::typeof(!), col_ind::Symbol)
   @ DataFrames ~/.julia/packages/DataFrames/58MUJ/src/dataframe/dataframe.jl:557
 [4] getproperty(df::DataFrame, col_ind::Symbol)
   @ DataFrames ~/.julia/packages/DataFrames/58MUJ/src/abstractdataframe/abstractdataframe.jl:431
 [5] top-level scope
   @ REPL[7]:1
```

or with `PythonCall`:

```julia-repl
julia> x.foo
ERROR: Python: AttributeError: 'list' object has no attribute 'foo'
Python stacktrace: none
Stacktrace:
 [1] pythrow()
   @ PythonCall.Core ~/.julia/packages/PythonCall/sQSpa/src/Core/err.jl:92
 [2] errcheck
   @ ~/.julia/packages/PythonCall/sQSpa/src/Core/err.jl:10 [inlined]
 [3] pygetattr(x::Py, k::String)
   @ PythonCall.Core ~/.julia/packages/PythonCall/sQSpa/src/Core/builtins.jl:62
 [4] getproperty(x::Py, k::Symbol)
   @ PythonCall.Core ~/.julia/packages/PythonCall/sQSpa/src/Core/Py.jl:273
 [5] top-level scope
   @ REPL[12]:1
```

I can't say I see much room there for automagic, if we provide a conventional error type though, we can give packages much more helpful error messages simply by opting-in to the standard error kind.

**Comment by arhik** [comment]
If we have `PropertyError` like in this draft #55298 would you guys prefer this PR work with `PropertyError` instead of `FieldError`. The draft I am sharing has following output. 

```julia
julia> struct D
               a::Float32
               b::Float64
       end

julia> d = D(1, 2)
D(1.0f0, 2.0)

julia> d.l
ERROR: PropertyError: instance of type `DataType` has no property `l`
DataType has following properties (:name, :super, :parameters, :types, :instance, :layout, :hash, :flags)

Stacktrace:
 [1] getproperty(x::D, f::Symbol)
   @ Base ./Base.jl:49
 [2] top-level scope
   @ REPL[3]:1

julia> d.a
1.0f0

julia> d.b
2.0

julia> D.l
ERROR: FieldError: type DataType has no field l
Stacktrace:
 [1] getproperty(x::Type, f::Symbol)
   @ Base ./Base.jl:43
 [2] top-level scope
   @ REPL[6]:1
 
 ```

**Comment by arhik** [comment]
> bump... the last comments seem more a discussion on future prospects, not issues with this PR.
> FieldError has the benefit of being automatically thrown in lots of cases, so adding a hint to it (this PR) would make many existing error messages better immediately.

True. It's more a discussion on future prospects. Not a blocking argument for this PR. Sorry If I am polluting this PR with digressions. I am just trying to check with others on a hypothetical question. `FieldError` treats Datatypes and objects the same. I was checking with other contributors if we would want to separate them. Separating them might help with more granularity and better error messages. I get the hint that @tecosaur might also share this opinion. 

> Test failures don't appear to be related, and also not sure whether failing "Labels / Check for blocking labels" means something I should do or not.

Yeah. Test failures are unrelated most likely. The label with pattern "needs *" is a blocking label. @KristofferC, @LilithHafner or anyone with rights will have to help you with that.

File: base/errorshow.jl
Comment (vtjnash) [inline base/errorshow.jl:1129]: New error hints should likely go in the `REPL` stdlib (where all the fuzzy string hints are located), not in Base

File: base/errorshow.jl
Comment (LilithHafner) [inline base/errorshow.jl:1126]:
```suggestion
```
I think we can simplify `_propertynames_bytype` by removing this branch. If the methods are the same, then `setdiff(props, fields)`, above, will be empty which has the same impact as returning `nothing` here.

Comment (LilithHafner) [reply base/errorshow.jl:1126]: That change request was for performance reasons, but the short circuit path is much slower than the default path. I don't think it should be accepted. Removing it results in 1.2-2x faster compile time and 10-15x faster runtime.

```julia-repl
x@x:~$ julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.4 (2024-06-04)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using Base: promote_op

julia> _extract_val(::Type{Val{V}}) where {V} = V
_extract_val (generic function with 1 method)

julia> _extract_val(@nospecialize _) = nothing
_extract_val (generic function with 2 methods)

julia> _propertynames_bytype(::Type{T}) where {T} =
           # which(propertynames, (T,)) === which(propertynames, (Any,)) ? nothing :
           _extract_val(promote_op(Val∘propertynames, T))
_propertynames_bytype (generic function with 1 method)

julia> @time @eval _propertynames_bytype(Dict)
  0.011929 seconds (25.30 k allocations: 1.731 MiB, 99.02% compilation time)
(:slots, :keys, :vals, :ndel, :count, :age, :idxfloor, :maxprobe)

julia> using Chairmarks

julia> @b Dict _propertynames_bytype
95.714 ns (0.03 allocs: 1.159 bytes)

julia> using LinearAlgebra

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.003214 seconds (15.16 k allocations: 1.025 MiB, 97.22% compilation time)
(:R, :Q, :p, :P)

julia> @b LinearAlgebra.QRPivoted _propertynames_bytype
179.878 ns (0.05 allocs: 1.561 bytes)

julia> 
x@x:~$ julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.4 (2024-06-04)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using Base: promote_op

julia> _extract_val(::Type{Val{V}}) where {V} = V
_extract_val (generic function with 1 method)

julia> _extract_val(@nospecialize _) = nothing
_extract_val (generic function with 2 methods)

julia> _propertynames_bytype(::Type{T}) where {T} =
           which(propertynames, (T,)) === which(propertynames, (Any,)) ? nothing :
           _extract_val(promote_op(Val∘propertynames, T))
_propertynames_bytype (generic function with 1 method)

julia> @time @eval _propertynames_bytype(Dict)
  0.014774 seconds (62.30 k allocations: 4.289 MiB, 99.48% compilation time)

julia> @b Dict _propertynames_bytype
1.471 μs (4.37 allocs: 233.263 bytes)

julia> using LinearAlgebra

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.006811 seconds (34.88 k allocations: 2.380 MiB, 98.59% compilation time)
(:R, :Q, :p, :P)

julia> using Chairmarks

julia> @b LinearAlgebra.QRPivoted _propertynames_bytype
1.781 μs (6.50 allocs: 368 bytes)
```

File: base/errorshow.jl
Comment (LilithHafner) [reply base/errorshow.jl]: > We can also make this much more efficient by adding an early exit to catch most cases

The early exit here _decreases_ performance. It's good to benchmark before recommending this sort of micro-optimization. See https://github.com/JuliaLang/julia/pull/55165#discussion_r1715275750 for benchmarks.

Comment (vtjnash) [reply base/errorshow.jl]: The benchmark result just means it doesn't have the other expected fixes compared to the code I suggested above. My version should compile infinitely faster, as it isn't compiled at all

File: base/errorshow.jl
Comment (KristofferC) [reply base/errorshow.jl:1129]: > New error hints should likely go in the REPL stdlib (where all the fuzzy string hints are located), not in Base

This doesn't make sense to me. Keep them all in the same place until they are wholesale moved.

File: base/errorshow.jl
Comment (LilithHafner) [reply base/errorshow.jl]: I modified the function you recommended by commenting out the line with `NTuple{Symbol}` which should only speed it up.

```julia-repl
julia> using Base: promote_op

julia> function _propertynames_bytype(T::Type)
           inferred_names = promote_op(Val∘propertynames, T)
           inferred_names isa DataType && inferred_names <: Val || return nothing
           inferred_names = inferred_names.parameters[1]
           inferred_names isa NTuple{Symbol} || return nothing
           return Symbol[inferred_names[i] for i in 1:length(inferred_names)]
       end
_propertynames_bytype (generic function with 4 methods)

julia> _propertynames_bytype(Dict)
ERROR: TypeError: in Vararg, in count, expected Int64, got Type{Symbol}
Stacktrace:
 [1] _propertynames_bytype(T::Type)
   @ Main ./REPL[8]:5
 [2] top-level scope
   @ REPL[9]:1
```

When I also add in your recommended optimization it is still _much_ slower than the PR's current implementation (and contains much more direct dependence on internals). I suspect this is be because the function you recommended executes more slowly than compilation.

```
x@x:~$ julia +nightly
A new `nightly` version is available. Install with `juliaup update`.
  o  | Version 1.12.0-DEV.1106 (2024-08-28)
 o o | Commit 644029282cf (2 days old master)
julia> _extract_val(::Type{Val{V}}) where {V} = V
_extract_val (generic function with 1 method)

julia> _extract_val(@nospecialize _) = nothing
_extract_val (generic function with 2 methods)

julia> _propertynames_bytype(::Type{T}) where {T} =
           _extract_val(promote_op(Val∘propertynames, T))
_propertynames_bytype (generic function with 1 method)

julia> using Base: promote_op

julia> @time @eval _propertynames_bytype(Dict)
  0.011136 seconds (20.98 k allocations: 1019.578 KiB, 66.72% compilation time)
(:slots, :keys, :vals, :ndel, :count, :age, :idxfloor, :maxprobe)

julia> using Chairmarks

julia> @b Dict _propertynames_bytype
98.244 ns (0.01 allocs: 0.589 bytes)

julia> using LinearAlgebra

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.002739 seconds (12.35 k allocations: 603.828 KiB, 96.23% compilation time)
(:R, :Q, :p, :P)

julia> @b LinearAlgebra.QRPivoted _propertynames_bytype
179.372 ns (0.02 allocs: 0.878 bytes)
x@x:~$ julia +nightly
A new `nightly` version is available. Install with `juliaup update`.
  o  | Version 1.12.0-DEV.1106 (2024-08-28)
 o o | Commit 644029282cf (2 days old master)
julia> function _propertynames_bytype(T::Type)
           which(propertynames, (T,)) === which(propertynames, (Any,)) && return nothing
           inferred_names = promote_op(Val∘propertynames, T)
           inferred_names isa DataType && inferred_names <: Val || return nothing
           inferred_names = inferred_names.parameters[1]
           #inferred_names isa NTuple{Symbol} || return nothing
           return Symbol[inferred_names[i] for i in 1:length(inferred_names)]
       end
_propertynames_bytype (generic function with 1 method)

julia> using Base: promote_op

julia> @time @eval _propertynames_bytype(Dict)
  0.286678 seconds (4.56 M allocations: 251.616 MiB, 21.61% gc time, 98.41% compilation time)

julia> using Chairmarks

julia> @b Dict _propertynames_bytype
2.326 μs (23.33 allocs: 985.333 bytes)

julia> using LinearAlgebra

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.032111 seconds (104.80 k allocations: 5.559 MiB, 99.51% compilation time: 18% of which was recompilation)
4-element Vector{Symbol}:
 :R
 :Q
 :p
 :P

julia> @b LinearAlgebra.QRPivoted _propertynames_bytype
5.844 μs (50 allocs: 2.512 KiB)
```

File: test/errorshow.jl
Comment (LilithHafner) [inline test/errorshow.jl:14]: These should be swapped
Right now hints looks funkey on `Dict` but I think swapping the hint order will fix it:
```
x@x:~$ juliaup add pr55165
Installing Julia pr55165-linux-aarch64
julia +pr55165
x@x:~$ julia +pr55165
  o  | Version 1.12.0-DEV.882 (2024-08-30)
 o o | aplavin:patch-15/4d52030318f (fork: 9 commits, 42 days)
julia> Dict().g
ERROR: FieldError: type Dict has no field `g`
Did you mean to access dict values using key: `:g` ? Consider using indexing syntax dict[:g]
, available fields: `slots`, `keys`, `vals`, `ndel`, `count`, `age`, `idxfloor`, `maxprobe`
Stacktrace:
 [1] getproperty(x::Dict{Any, Any}, f::Symbol)
   @ Base ./Base.jl:49
 [2] top-level scope
   @ REPL[1]:1
```
```suggestion
Base.Experimental.register_error_hint(Base.fielderror_listfields_hint_handler, FieldError)
Base.Experimental.register_error_hint(Base.fielderror_dict_hint_handler, FieldError)
```

**Comment by LilithHafner** [comment]
Aside from the Dict hint ordering, this LGTM

**Review by LilithHafner — APPROVED** [review]

**Comment by LilithHafner** [comment]
I'm going to merge this in a couple of days unless @vtjnash objects.

It's a UX improvement that does not impact API and seems pretty low-risk to me.

File: base/errorshow.jl
Comment (vtjnash) [reply base/errorshow.jl]: Something seems vaguely suspicious, since that is quite a bit of extra allocation. Especially just noting that my function is type-stable, while yours is testing the performance of only a single signature and re-specializes on every different length of number of fields.
```
julia> @time @eval _propertynames_bytype(Dict)
  0.622177 seconds (4.69 M allocations: 262.580 MiB, 16.18% gc time, 98.57% compilation time)

julia> using LinearAlgebra

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.088465 seconds (108.28 k allocations: 5.823 MiB, 99.58% compilation time: 22% of which was recompilation)
4-element Vector{Symbol}:
 :R
 :Q
 :p
 :P

julia> @time @eval _propertynames_bytype(LinearAlgebra.QRPivoted)
  0.000366 seconds (116 allocations: 5.094 KiB, 0.43% compilation time: 100% of which was recompilation)
4-element Vector{Symbol}:
 :R
 :Q
 :p
 :P
```

It appears that the actual issue is that `which` hasn't been compiled yet, so it takes a bit of extra time:
```
julia> which(Int, (Int,));

julia> @time @eval _propertynames_bytype(Dict)
  0.057656 seconds (65.06 k allocations: 3.342 MiB, 83.99% compilation time)
```

So whereas your version takes 11ms to return a new query, my version takes on the order of 0.4ms

Comment (LilithHafner) [reply base/errorshow.jl]: You benchmarked it taking 0.4ms to rerun the _same_ query, not a new query. In any case, the PR's version is faster on the usecase you posted. Would you please post a reproducible benchmark demonstrating a use case where your proposal is faster than the PR's implementation?

Comment (vtjnash) [reply base/errorshow.jl]: It is 0.4ms to run any query that doesn't specialize the `which` method or to rerun any query, as it doesn't run inference for either of those cases. The current PR version is much slower than the former and leaks memory, despite that being the most common case.

File: base/errorshow.jl
Comment (LilithHafner) [inline base/errorshow.jl:1128]: `NTuple{Symbol}` is not a thing.
```suggestion
    inferred_names isa NTuple{<:Any, Symbol} || return nothing
```

Comment (nsajko) [reply base/errorshow.jl]: FWIW IMO `Tuple{Vararg{Symbol}}` is nicer than `NTuple{<:Any, Symbol}`.

**Comment by LilithHafner** [comment]
Thank you for your persistence, @aplavin. Sorry for putting you in the middle of conflicting feedback from different maintainers and thank you for being willing to make and remake requested changes. This is a nice UX improvement and I'm glad to have it.

**Pull request merged**
